docs/writing-benchmarks.md

```diff
@@ -0,0 +1,78 @@
+    }
+
+    @Benchmark
+    fun benchmarkMethod(): Int {
+        return list.sum()
+    }
+
+    @TearDown
+    fun cleanup() {
+        list.clear()
+    }
+}
+```
+
+### Explaining the Annotations
+
+#### @State
+
+The `@State` annotation, when set to `Scope.Benchmark`, is applied to a class to represent that this class is responsible for holding the state or data for your benchmark tests. This class instance is shared among all benchmark threads, creating a consistent environment across all warmups, iterations, and measured executions. It helps in managing stateful operations in a multi-threaded context, providing an efficient way of handling state that's thread-safe and consistent across multiple runs. This ensures accurate and reliable benchmarking results, as the shared state remains the same throughout all the tests, preventing any discrepancies that could affect the outcomes. In the Kotlin/JVM and Java taregts, apart from `Scope.Benchmark`, you also have access to `Scope.Thread` and `Scope.Group`. `Scope.Thread` ensures each thread has its own unique state instance, while `Scope.Group` allows state sharing within a benchmark thread group. However, for other target environments, only `Scope.Benchmark` is supported, limiting the scope options to a single instance that is shared among all threads. In our snippet, the ExampleBenchmark class uses the @State(Scope.Benchmark) annotation, indicating that the state in this class is shared across all benchmark threads.
+
+#### @Setup
+
+The `@Setup` annotation is used to mark a method that sets up the necessary preconditions for your benchmark test. It serves as a preparatory step where you set up the environment for the benchmark, performing tasks such as generating data, establishing database connections, or preparing any other resources your benchmark requires. In the Kotlin/JVM and Java targets, the `@Setup` annotation can operate on two levels - `Iteration` and `Trial`. In `Iteration` level, the setup method is run before each benchmark iteration, ensuring a consistent state across all iterations. On the other hand, `Trial` level execution runs the setup method once for the entire set of benchmark method iterations, suitable when state modifications are part of the benchmark itself. Specify the level using `@Setup(Level.Trial)` or `@Setup(Level.Iteration)`; if not defined, it defaults to `Level.Trial`. The key point to remember is that the `@Setup` method's execution time is not included in the final benchmark results - the timer starts only when the `@Benchmark` method begins. This makes `@Setup` an ideal place for initialization tasks that should not impact the timing results of your benchmark. By using the `@Setup` annotation, you ensure consistency across all executions of your benchmark, providing accurate and reliable results. In the provided example, the `@Setup` annotation is used to prepare a list containing integers from 0 to a specified parameter value.
```

Comment:
On what level does the `Setup` operate in other targets?

Reply:
Setup defaults to @Level.Trial on all targets as you noted earlier. I will update the documentation to reflect this info. https://github.com/openjdk/jmh/blob/09c78d5d0752ffc409e64ca4cabe9dc7b96704d7/jmh-core/src/main/java/org/openjdk/jmh/annotations/Setup.java